server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java

```diff
@@ -172,52 +174,64 @@ private void confirmDeletes(GarbageCollectionEnvironment gce,
+          String cq = key.getColumnQualifier().toString();
+
+          String reference = cq;
+          if (cq.startsWith("/")) {
+            String tableID = new String(KeyExtent.tableOfMetadataRow(key.getRow()));
+            reference = "/" + tableID + cq;
+          } else if (!cq.contains(":") && !cq.startsWith("../")) {
+            throw new RuntimeException("Bad file reference " + cq);
+          }
 
-        // WARNING: This line is EXTREMELY IMPORTANT.
-        // You MUST REMOVE candidates that are still in use
-        if (candidateMap.remove(reference) != null)
-          log.debug("Candidate was still in use: " + reference);
+          reference = makeRelative(reference, 3);
 
-        String dir = reference.substring(0, reference.lastIndexOf('/'));
-        if (candidateMap.remove(dir) != null)
-          log.debug("Candidate was still in use: " + reference);
+          // WARNING: This line is EXTREMELY IMPORTANT.
+          // You MUST REMOVE candidates that are still in use
+          if (candidateMap.remove(reference) != null)
+            log.debug("Candidate was still in use: " + reference);
 
-      } else if (TabletsSection.ServerColumnFamily.DIRECTORY_COLUMN.hasColumns(key)) {
-        String tableID = new String(KeyExtent.tableOfMetadataRow(key.getRow()));
-        String dir = entry.getValue().toString();
-        if (!dir.contains(":")) {
-          if (!dir.startsWith("/"))
-            throw new RuntimeException("Bad directory " + dir);
-          dir = "/" + tableID + dir;
-        }
+          String dir = reference.substring(0, reference.lastIndexOf('/'));
+          if (candidateMap.remove(dir) != null)
+            log.debug("Candidate was still in use: " + reference);
 
-        dir = makeRelative(dir, 2);
+        } else if (TabletsSection.ServerColumnFamily.DIRECTORY_COLUMN.hasColumns(key)) {
+          readTracker.markDirSeen();
+          String tableID = new String(KeyExtent.tableOfMetadataRow(key.getRow()));
+          String dir = entry.getValue().toString();
+          if (!dir.contains(":")) {
+            if (!dir.startsWith("/"))
+              throw new RuntimeException("Bad directory " + dir);
+            dir = "/" + tableID + dir;
+          }
 
-        if (candidateMap.remove(dir) != null)
-          log.debug("Candidate was still in use: " + dir);
-      } else
-        throw new RuntimeException(
-            "Scanner over metadata table returned unexpected column : " + entry.getKey());
-    }
+          dir = makeRelative(dir, 2);
 
+          if (candidateMap.remove(dir) != null) {
+            log.debug("Candidate was still in use: " + dir);
+          }
+        } else if (TabletsSection.TabletColumnFamily.PREV_ROW_COLUMN.hasColumns(key)) {
+          readTracker.markPrevRowSeen();
+        } else
+          throw new RuntimeException(
+              "Scanner over metadata table returned unexpected column : " + entry.getKey());
+      }
+    } // closed tracker - checks last row.
```

Comment:
```suggestion
      // the tracker is closed at the end of this block; it will check the last row
      // in its close method as its final action
    }
```